### DISCUSSION TITLE
How can I remove the influence of the port labels on the source/target positions?

### DISCUSSION THREAD

**Discussion by WaldoJr24** [clientIO/joint#2793] [Q&A] [answered]
### Introduction

If i connect two ports with long labels, the links always have a small gap to the port itself.
So it depends on the port label where the source/target position is.

I use the AvoidRouter to route the links.

Do you know how I can leave the labels out of the bbox calculation?
Thanks!

![384419175-c47f77b9-14dd-4b73-8042-f4186551b8fc](https://github.com/user-attachments/assets/bb1d5712-6109-490e-961b-4e1091fae095)

### Steps to reproduce

Just connect the ports with longer labels

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
The link needs to connect the `<circle/>` (not the whole `<g/>` that includes also the `<text/>`) i.e.
the [`magnet`](https://docs.jointjs.com/api/dia/attributes/#magnet) property needs to be set on the port's circle.

The `AvoidRouter` does not play any role here.

**Reply by WaldoJr24** [reply]
OK sorry I thought it because it occurred the first time i changed to AvoidRouter.


I think I'm already connecting to the circle itself:

console log of the sourceMagnet  on validateConnection:
sourceMagnet: "<circle id="v-112" joint-selector="portBody" magnet="true" r="8" cy="3" fill="green" stroke="grey" stroke-width="1" filter="url(#dropShadowv-2862351534)"">

![image](https://github.com/user-attachments/assets/2e36f3de-5414-4052-89cc-5d3213335e74)



But the link itself just goes to the <g>:
![image](https://github.com/user-attachments/assets/d73854ba-671b-41b5-8cc8-8d14513f73ec)



On start of the linking process it looks like this (on correct position):
![image](https://github.com/user-attachments/assets/fef7db0d-0210-42f2-b7db-ac1784d1a849)


After the link is set and gets routed it looks like this:
![image](https://github.com/user-attachments/assets/2d56cf6c-0fbb-4f0d-a958-f6ad9f27d7c0)

**Reply by kumilingus** [reply]
Do you add the green link programmatically? (I can see there is a black link during dragging).

The link's `source` / `target` should have `magnet` property.
```ts
{
  id: 'e9c79cab-7201-420c-89fb-319b77a27e20',
  magnet: 'portBody',
  port: '9741ea0b-47bb-47dc-b42e-42eac648b193'
}
```

Alternatively, you could use [`magentSelector`](https://docs.jointjs.com/api/dia/attributes/#magnetselector) attribute. See https://jsfiddle.net/kumilingus/7nhwxL03/.

**Reply by WaldoJr24** [reply, marked as answer]
Yes, I'm setting the link color dynamically based on the port group in the `link:connect` event.


I found a working solution:

First I added magnet: 'portBody' to the source and target of the link, which seemed to work at first. However, after rerouting the link, the positions snapped back to the <g> element.

To resolve this, I added `magnet: 'portBody'` to the linkAttributes inside the `routeLink()` function of the AvoidRouter script:

```
const linkAttributes: { [key: string]: any } = {
            source: {
                id: sourceId,
                magnet: 'portBody',
                port: sourcePortId || null,
                anchor: {
                    name: 'modelCenter',
                },
            },
            target: {
                id: targetId,
                magnet: 'portBody',
                port: targetPortId || null,
                anchor: {
                    name: 'modelCenter',
                },
            },
        };
```

 This change appears to have fixed the issue, and everything is working as expected now.
